File: 2024/March/3-6/LinkedListCycle.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        fast = head
        slow = head
        loops = False

        while not loops:
            if fast is None:
                break
            
            fast = fast.next
            if fast is None:
                break
            fast = fast.next
            slow = slow.next
            if fast is None:
                break
            if fast == slow:
                loops = True
        
        return loops
        
        

File: 2024/March/3-6/LinkedListCycle.java

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode fast = head;   // Fast incrementing pointer
        ListNode slow = head;   // Slow incrementing pointer
        boolean loops = false;  // Loops return val

        while(!loops) {
            if(fast == null) {  // If fast ever is null, there is no loop
                break;
            }

            fast = fast.next;
            if(fast == null) {
                break;
            }
            fast = fast.next;
            slow = slow.next;
            if(fast == null) {
                break;
            }
            if(fast == slow) {  // If fast catches up to slow, there must be a loop
                loops = true;
            }
        }

        return loops;
    }
}